Document using custom Redis commands (#18463)
quarkusio/quarkus: docs/src/main/asciidoc/redis-reference.adoc

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
////
= Redis Reference Guide

:extension-status: preview
include::./attributes.adoc[]
:numbered:
:sectnums:

[[custom_redis_commands]]
== How to use custom Redis Commands

As the list of commands which are supported out-of-box by both https://github.com/quarkusio/quarkus/blob/main/extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/RedisClient.java[`RedisClient`] and https://github.com/quarkusio/quarkus/blob/main/extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/reactive/ReactiveRedisClient.java[`ReactiveRedisClient`] depends on what is available in https://github.com/vert-x3/vertx-redis-client[`vertx-redis-client`], then there might be case when you need a command which is not (yet) available via https://github.com/vert-x3/vertx-redis-client[`vertx-redis-client`].

See:

https://github.com/quarkusio/quarkus/issues/18351

> https://github.com/vert-x3/vertx-redis-client/issues/299

Reviewer: Not sure it worths mentioning the issues here.

Author: I can remove these, as I wanted to mention what were the initial reasons why this was needed.

Reviewer: Maybe as a note admonition?

Author: Yes, I think it will be better - as then someone could see the changes in a "bigger picture".

Reviewer 2: Frankly, I would just drop them, they don't really bring much to the plate except noise.

Author: OK, I finally removed them.

> In such case (if you don't want to wait for the new command to appear in https://github.com/vert-x3/vertx-redis-client[`vertx-redis-client`]), you can implement it directly in either https://github.com/quarkusio/quarkus/blob/main/extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/RedisClient.java[`RedisClient`] or https://github.com/quarkusio/quarkus/blob/main/extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/reactive/ReactiveRedisClient.java[`ReactiveRedisClient`]. In order to do so, you will need:

Reviewer: ```suggestion
In such case (if you don't want to wait for the new command to be supported in https://github.com/vert-x3/vertx-redis-client[`vertx-redis-client`]), you can implement it in either https://github.com/quarkusio/quarkus/blob/main/extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/RedisClient.java[`RedisClient`] or https://github.com/quarkusio/quarkus/blob/main/extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/reactive/ReactiveRedisClient.java[`ReactiveRedisClient`]. 
In order to do so, you will need:
```

[source,java]
----
@Override
public Response zunion(List<String> args) {
    final io.vertx.mutiny.redis.client.Command ZUNION = Command.create("zunion", -3, 0, 0, 0, false, true, true, false);
    final io.vertx.mutiny.redis.client.Request requestWithArgs = args.stream().reduce(
            io.vertx.mutiny.redis.client.Request.cmd(ZUNION),
            (request, s) -> request.arg(s),
            (request, request2) -> request);

    return await(mutinyRedis.send(requestWithArgs));
}
----

- Updating the https://github.com/quarkusio/quarkus/blob/main/extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/reactive/ReactiveRedisClient.java[`ReactiveRedisClient`] interface, i.e.:

[source,java]
----
Uni<Response> zunion(List<String> args);

Response zunionAndAwait(List<String> args);
----

- Updating the https://github.com/quarkusio/quarkus/blob/main/extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/runtime/ReactiveRedisClientImpl.java[`ReactiveRedisClientImpl`], i.e.:

> 

Reviewer 3: This is really good reference content. But it covers adding the missing commands on the extension which is good but the downside to the approach is that it might require a new release of the extension (Quarkus release) before the user can access the command.

Can we add a section on how the user can add the code new command on their own application? The idea will be the same but this time injecting the `MutinyRedis` client into their application code.

Author: Yes, I see.

The description was initially meant to describe how it's possible to add a specific command to `quarkus-redis-client` without waiting for the changes in `vertx-redis-client` - which of course requires a release - as you commented.

I can further describe how this can be done without changes in the `quarkus-redis-client` extension, so this will be more "developer friendly".

[NOTE]
====
Please see these issues for the initial idea:

https://github.com/quarkusio/quarkus/issues/18351

https://github.com/vert-x3/vertx-redis-client/issues/299
====

In such case (if you don't want to wait for the new command to be supported in https://github.com/vert-x3/vertx-redis-client[`vertx-redis-client`]), you can implement it in either https://github.com/quarkusio/quarkus/blob/main/extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/RedisClient.java[`RedisClient`] or https://github.com/quarkusio/quarkus/blob/main/extensions/redis-client/runtime/src/main/java/io/quarkus/redis/client/reactive/ReactiveRedisClient.java[`ReactiveRedisClient`]. 
In order to do so, you will need:

> - Generate a new `Command` based on the nodejs code available in https://github.com/vert-x3/vertx-redis-client[`vertx-redis-client`] repository:

Reviewer 2: Having a list with only one item is weird. Maybe just `need to ...`.

Author: @gsmet Actually it's not a single item list, but five items:

- Generate a new `Command` based on the nodejs code ...
- Updating the RedisClient ...
- Updating the RedisClientImpl ...
- Updating the ReactiveRedisClient ...
- Updating the ReactiveRedisClientImpl ...

The problem is that the 1st bullet is quite extensive, so it might confuse readers.

> If you don't have a redis service running locally, you can run redis in a docker container:

Reviewer 2: ```suggestion
If you don't have a Redis service running locally, you can run Redis in a Docker container:
```

[source,shell script]
----
docker run --name redis -p 7006:6379 -d redis
----

Next, while being in https://github.com/vert-x3/vertx-redis-client[`vertx-redis-client`] root folder execute:

[source,shell script]
----
cd tools
npm i
npm start
----

> The above sequence of commands should update the https://github.com/vert-x3/vertx-redis-client/blob/master/src/main/java/io/vertx/redis/client/Command.java[`Command.java`] file, so it includes all the possible commands supported by a particular redis version.

Reviewer 2: ```suggestion
The above sequence of commands should update the https://github.com/vert-x3/vertx-redis-client/blob/master/src/main/java/io/vertx/redis/client/Command.java[`Command.java`] file, so it includes all the possible commands supported by a particular Redis version.
```

[source,java]
----
Command ZUNION = Command.create("zunion", -3, 0, 0, 0, false, true, true, false);
----

This definition is very important as we will have to use it in the service.

> Once we have this Command we can start to update the redis-client extension by:

Reviewer 2: ```suggestion
Once we have this `Command` we can start to update the redis-client extension by:
```
